### ISSUE TITLE
Using Mio for polling crossbeam channels?

### ISSUE THREAD

**Issue by JonForShort** [tokio-rs/mio#1607]
Hello, thank you for the project.  I am working on a small program where I need to poll on file descriptors, sockets and crossbeam channels.  I am currently using Mio for polling on file descriptors and sockets using similar code as this example (https://docs.rs/mio/0.8.4/mio/struct.Poll.html).

However for crossbeam channels (https://docs.rs/crossbeam-channel/latest/crossbeam_channel/), I have to use their ```select``` API (https://docs.rs/crossbeam-channel/0.5.6/crossbeam_channel/struct.Select.html).  Having to use two different mechanisms for polling has made my code a bit messy.

I was just curious if there is any current ways to Mio's poll mechanism on crossbeam channels? Thanks!

**Comment by Thomasdezeeuw** [comment]
Hello @JonForShort Mio used to have support for user space events, but it was removed from the crate to focus on providing a smaller API. So, to answer your question

> I was just curious if there is any current ways to Mio's poll mechanism on crossbeam channels? Thanks!

I'm afraid not.

I've personally implemented this using a wrapper around the crossbeam channel using [`mio::Waker`](https://docs.rs/mio/latest/mio/struct.Waker.html). Basically every time you send a message into the channel call `mio::Waker::wake` to wake up the other side in case it's polling. Then remove all messages from the channel before calling `mio::Poll::poll`. However if you send a lot of messages the overhead of a system call might be too much. For that I've created [`ThreadWaker`](https://github.com/Thomasdezeeuw/heph/blob/d44eba4650c928fecfbde2eb68814d640af9da4c/rt/src/thread_waker.rs#L10) in my own project, which only makes a system call when the other side is actually polling, see the usage here: https://github.com/Thomasdezeeuw/heph/blob/d44eba4650c928fecfbde2eb68814d640af9da4c/rt/src/worker.rs#L585-L634.

If you have any more questions let us know.
